precompile/contracts/feemanager/simulated_test.go

```diff
@@ -0,0 +1,314 @@
+
+	type testCase struct {
+		name string
+		test func(t *testing.T, backend *sim.Backend, feeManagerIntf *feemanagerbindings.IFeeManager)
+	}
+
+	testCases := []testCase{
+		{
+			name: "should verify admin has admin role",
+			test: func(t *testing.T, _ *sim.Backend, feeManager *feemanagerbindings.IFeeManager) {
+				allowlisttest.VerifyRole(t, feeManager, adminAddress, allowlist.AdminRole)
+			},
+		},
+		{
+			name: "should verify new contract has no role",
+			test: func(t *testing.T, backend *sim.Backend, feeManager *feemanagerbindings.IFeeManager) {
+				testContractAddr, _ := deployFeeManagerTest(t, backend, admin)
+				allowlisttest.VerifyRole(t, feeManager, testContractAddr, allowlist.NoRole)
+			},
+		},
+		{
+			name: "contract should not be able to change fee without enabled",
+			test: func(t *testing.T, backend *sim.Backend, feeManager *feemanagerbindings.IFeeManager) {
+				testContractAddr, testContract := deployFeeManagerTest(t, backend, admin)
+
+				allowlisttest.VerifyRole(t, feeManager, adminAddress, allowlist.AdminRole)
+				allowlisttest.VerifyRole(t, feeManager, testContractAddr, allowlist.NoRole)
+
+				// Try to change fee config from contract without being enabled - should fail during gas estimation
+				_, err := testContract.SetFeeConfig(
+					admin,
+					cchainFeeConfig.GasLimit,
+					big.NewInt(int64(cchainFeeConfig.TargetBlockRate)),
+					cchainFeeConfig.MinBaseFee,
+					cchainFeeConfig.TargetGas,
+					cchainFeeConfig.BaseFeeChangeDenominator,
+					cchainFeeConfig.MinBlockGasCost,
+					cchainFeeConfig.MaxBlockGasCost,
+					cchainFeeConfig.BlockGasCostStep,
+				)
+				require.ErrorContains(t, err, "execution reverted")
+			},
+		},
+		{
+			name: "contract should be added to manager list",
+			test: func(t *testing.T, backend *sim.Backend, feeManager *feemanagerbindings.IFeeManager) {
+				testContractAddr, _ := deployFeeManagerTest(t, backend, admin)
+				allowlisttest.VerifyRole(t, feeManager, adminAddress, allowlist.AdminRole)
+				allowlisttest.VerifyRole(t, feeManager, testContractAddr, allowlist.NoRole)
+				allowlisttest.SetAsEnabled(t, backend, feeManager, admin, testContractAddr)
+				allowlisttest.VerifyRole(t, feeManager, testContractAddr, allowlist.EnabledRole)
+			},
+		},
+		{
+			name: "admin should be able to change fees",
+			test: func(t *testing.T, backend *sim.Backend, feeManager *feemanagerbindings.IFeeManager) {
+				testContractAddr, testContract := deployFeeManagerTest(t, backend, admin)
+				allowlisttest.SetAsEnabled(t, backend, feeManager, admin, testContractAddr)
+
+				verifyFeeConfigMatches(t, testContract, genesisFeeConfig)
+				setFeeConfig(t, backend, testContract, admin, cchainFeeConfig)
+				verifyFeeConfigMatches(t, testContract, cchainFeeConfig)
+
+				// Verify last changed at block number
+				lastChangedAt, err := testContract.GetFeeConfigLastChangedAt(nil)
+				require.NoError(t, err)
+				require.NotZero(t, lastChangedAt.Uint64())
```

Comment:
can we actually check if this corresponds to the exact expected block number?

Reply:
Yup!